src/boardState.ts

```diff
@@ -0,0 +1,239 @@
+
+        while(numberUnresolved >0){
+
+            let firstUnresolved:ShotImpliedPlacements;
+            let index: number =0;
+
+            while(!firstUnresolved){
+                if(shotImpliedPlacementsList[index].resolved == false){
+                    firstUnresolved = shotImpliedPlacementsList[index];
+                }
+                index++;
+            }
+            
+            let nextShip:ShipPlacement
+            if(firstUnresolved.allValidPlacements.length>0){
+                if(!exhaustive){
+                    nextShip = firstUnresolved.pickRandomPlacement();
+                }
+                else{
+                    let nextStep: number = 0;
+                    if(pathToFollow.length > currentPathIndex){ //if the length is good
+                        nextStep = pathToFollow[currentPathIndex];
+                        if(nextStep > firstUnresolved.allValidPlacements.length-1){
+                            let nextPath:number[] = currentPath;
+                            nextPath[nextPath.length-1]++;
+                            return {state: new BoardState([]),found:false, nextPath: nextPath};
+                        }
+                    }
+                    currentPathIndex++;
+                    currentPath.push(nextStep);
+                    nextShip = firstUnresolved.allValidPlacements[nextStep]
+                }
+            }
+            else{
+                if(!exhaustive)
+                {
+                    return {state: new BoardState([]), found:false, nextPath: []};
+                }else{
+                    let nextPath:number[] = currentPath;
+                    nextPath[nextPath.length-1]++;
+                    return {state: new BoardState([]),found:false, nextPath: nextPath};
+                }
+            }
+
+            let occupiedPositions: Array<Position> = nextShip.getOccupiedPositions();
+
+            possibleShips.push(nextShip);
+            numberRemaining[nextShip.length]--;
+            for(var i=0;i<validShipPositions.length; i++){
+                validShipPositions[i].removePossibilities(nextShip);
+            }
+            
+            for(var j=0; j<shotImpliedPlacementsList.length;j++){
+                for(var i=0; i<occupiedPositions.length;i++){
+                    
+                    if((!shotImpliedPlacementsList[j].resolved)&& Position.isEqual(shotImpliedPlacementsList[j].shot.Position, occupiedPositions[i])){
+                        
+                        numberUnresolved--;
+                        shotImpliedPlacementsList[j].resolved = true;
+                    }
+                    
+                }
+                shotImpliedPlacementsList[j].updateRemainingShips(numberRemaining);
+                shotImpliedPlacementsList[j].removeInvalidPositions(validShipPositions);
+            }
+
+            shotImpliedPlacementsList.sort((a,b,)=>{return a.allValidPlacements.length - b.allValidPlacements.length});
+        }
```

Comment:
There's a huge amount in this while loop. Could it not be separated into something like:

```
while(numberUnresolved >0) {
  let firstUnresolved = this.getFirstUnresolvedShot(params);
  this.checkValidShotPlacements(firstUnresolved, otherParams);
  etc.
}
```

...it would make this a lot easier to follow.